File: designpatterns/src/main/java/be/ecam/pattern/concurrency/ProducerConsumer.java
package be.ecam.pattern.concurrency;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public class ProducerConsumer {
    private static final BlockingQueue<Integer> buffer = new ArrayBlockingQueue<>(10);
//    private static final Object MONITOR = new Object();

    private static final Thread producer = new Thread(() -> {
        for (int i = 0; i < 10; i++) {
            try {
                buffer.put(i);
//                synchronized (MONITOR) {
                    System.out.println("producing " + i);
//                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    });

    private static final Thread consumer = new Thread(() -> {
        while (true) {
            try {
                int take = buffer.take();
//                synchronized (MONITOR) {
                    System.out.println("consuming " + take);
//                }
                if (take == 9) {
                    break;
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    });

    public static void main(String[] args) {
        producer.start();
        consumer.start();
    }
}

File: designpatterns/src/main/java/be/ecam/pattern/concurrency/ProducerConsumer.kt
package be.ecam.pattern.concurrency

import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel

val channel = Channel<Int>()

fun main() = runBlocking {
    val producerJob = launch { // launch a coroutine to act as a producer
        producer()
    }

    val consumerJob = launch { // launch a coroutine to act as a consumer
        consumer()
    }

    // Ensure that the main function does not exit until both producer and consumer coroutines are done
    producerJob.join()
    consumerJob.join()
}

suspend fun producer() {
    repeat(5) {
        delay(1000) // simulate some work
        val data = it + 1
        println("Producing $data")
        channel.send(data)
    }
    channel.close() // close the channel when done producing
}

suspend fun consumer() {
    for (data in channel) {
        println("Consuming $data")
    }
}